File: src/main/java/com/wxl/dyttcrawler/core/ExecutorThreadPool.java
package com.wxl.dyttcrawler.core;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import org.springframework.util.Assert;

import java.util.concurrent.Future;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Create by wuxingle on 2020/6/26
 * ExecutorService包装
 */
public class ExecutorThreadPool implements ThreadPool {

    private static final int MAX_THREAD_NUM = 200;

    private ThreadPoolExecutor threadPoolExecutor;

    private int maxThreadNum;

    private AtomicInteger activeThreadNum = new AtomicInteger();

    public ExecutorThreadPool(int maxThreadNum) {
        Assert.isTrue(maxThreadNum > 0 && maxThreadNum <= MAX_THREAD_NUM, "max thread must in (0,200]");
        this.maxThreadNum = maxThreadNum;
        // cache pool
        this.threadPoolExecutor = new ThreadPoolExecutor(0, Integer.MAX_VALUE,
                10, TimeUnit.MINUTES,
                new SynchronousQueue<>(),
                new ThreadFactoryBuilder().setDaemon(true).setNameFormat("crawler-pool-%s").build(),
                new ThreadPoolExecutor.AbortPolicy()
        );
    }

    /**
     * 任务提交
     */
    @Override
    public synchronized Future<?> execute(Runnable runnable) throws InterruptedException {
        while (activeThreadNum.get() >= maxThreadNum) {
            this.wait();
        }
        activeThreadNum.incrementAndGet();
        return threadPoolExecutor.submit(() -> {
            try {
                runnable.run();
            } finally {
                synchronized (ExecutorThreadPool.this) {
                    activeThreadNum.decrementAndGet();
                    ExecutorThreadPool.this.notify();
                }
            }
        });
    }

    /**
     * 关闭
     */
    @Override
    public void shutdown() throws InterruptedException {
        threadPoolExecutor.shutdown();
        while (!threadPoolExecutor.isTerminated()) {
            threadPoolExecutor.awaitTermination(5, TimeUnit.SECONDS);
        }
    }

    /**
     * 是否关闭
     */
    @Override
    public boolean isShutDown() {
        return threadPoolExecutor.isShutdown();
    }

    /**
     * 运行中线程
     */
    @Override
    public int getActiveThreadNum() {
        return activeThreadNum.get();
    }

    /**
     * 总线程
     */
    @Override
    public int getMaxThreadNum() {
        return maxThreadNum;
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/core/ExecutorThreadPool.kt
package com.wxl.dyttcrawler.core

import com.google.common.util.concurrent.ThreadFactoryBuilder
import java.util.concurrent.Future
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock

/**
 * Create by wuxingle on 2021/10/08
 * ExecutorService包装
 * todo
 */
class ExecutorThreadPool(override val maxThreadNum: Int) : ThreadPool {

    private val threadPoolExecutor = ThreadPoolExecutor(
        0, Integer.MAX_VALUE,
        10, TimeUnit.MINUTES,
        SynchronousQueue(),
        ThreadFactoryBuilder().setDaemon(true).setNameFormat("crawler-pool-%s").build(),
        ThreadPoolExecutor.AbortPolicy()
    )

    private val activeThread = AtomicInteger()

    private val lock = ReentrantLock()

    private val condition = lock.newCondition()

    /**
     * 任务提交
     * 阻塞到提交成功
     */
    override fun execute(runnable: Runnable): Future<*> {
        lock.lock()
        try {
            while (activeThread.get() >= maxThreadNum) {
                condition.await()
            }
            activeThread.incrementAndGet()
            return threadPoolExecutor.submit {
                try {
                    runnable.run()
                } finally {
                    lock.lock()
                    try {
                        activeThread.decrementAndGet()
                        condition.signal()
                    } finally {
                        lock.unlock()
                    }
                }
            }
        } finally {
            lock.unlock()
        }
    }

    /**
     * 关闭
     * 阻塞到关闭结束
     */
    override fun shutdown() {
        threadPoolExecutor.shutdown()
        while (!threadPoolExecutor.isTerminated) {
            threadPoolExecutor.awaitTermination(5, TimeUnit.SECONDS)
        }
    }

    /**
     * 是否关闭
     */
    override val isShutDown: Boolean
        get() = threadPoolExecutor.isShutdown

    /**
     * 运行中线程数
     */
    override val activeThreadNum: Int
        get() = activeThread.get()
}
